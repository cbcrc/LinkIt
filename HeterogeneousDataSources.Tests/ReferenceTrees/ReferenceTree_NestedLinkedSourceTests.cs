using ApprovalTests.Reporters;
using HeterogeneousDataSources.Tests.Shared;
using NUnit.Framework;
using RC.Testing;

namespace HeterogeneousDataSources.Tests {
    [UseReporter(typeof(DiffReporter))]
    [TestFixture]
    public class ReferenceTree_NestedLinkedSourceTests
    {
        [Test]
        public void CreateRootReferenceTree()
        {
            var loadLinkProtocolBuilder = new LoadLinkProtocolBuilder();
            loadLinkProtocolBuilder.For<LinkedSource>()
                .LoadLinkReference(
                    linkedSource => linkedSource.Model.PreImageId,
                    linkedSource => linkedSource.PreImage
                )
                .LoadLinkNestedLinkedSource(
                    linkedSource => linkedSource.Model.PersonId,
                    linkedSource => linkedSource.Person
                )
                .LoadLinkReference(
                    linkedSource => linkedSource.Model.PostImageId,
                    linkedSource => linkedSource.PostImage
                );
            loadLinkProtocolBuilder.For<PersonLinkedSource>()
                .LoadLinkReference(
                    linkedSource => linkedSource.Model.SummaryImageId,
                    linkedSource => linkedSource.SummaryImage
                );
            var sut = new LoadLinkConfig(loadLinkProtocolBuilder.GetLoadLinkExpressions());

            var actual = sut.CreateRootReferenceTree<LinkedSource>();

            ApprovalsExt.VerifyPublicProperties(actual);
        }

        public class LinkedSource : ILinkedSource<Model> {
            public Model Model { get; set; }
            public Image PreImage { get; set; }
            public PersonLinkedSource Person { get; set; }
            public Image PostImage { get; set; }
        }

        public class Model {
            public int Id { get; set; }
            public string PreImageId { get; set; }
            public string PersonId { get; set; }
            public string PostImageId { get; set; }
        }
    }
}
