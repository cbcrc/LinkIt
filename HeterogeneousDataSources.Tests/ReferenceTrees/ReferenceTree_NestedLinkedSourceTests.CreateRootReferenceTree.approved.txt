{
  "node": {
    "referenceType": "HeterogeneousDataSources.Tests.ReferenceTree_NestedLinkedSourceTests+Model, HeterogeneousDataSources.Tests, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null",
    "linkTargetId": "root"
  },
  "children": [
    {
      "node": {
        "referenceType": "HeterogeneousDataSources.Tests.Shared.Image, HeterogeneousDataSources.Tests, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null",
        "linkTargetId": "HeterogeneousDataSources.Tests.ReferenceTree_NestedLinkedSourceTests+LinkedSource/PreImage"
      },
      "children": []
    },
    {
      "node": {
        "referenceType": "HeterogeneousDataSources.Tests.Shared.Person, HeterogeneousDataSources.Tests, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null",
        "linkTargetId": "HeterogeneousDataSources.Tests.ReferenceTree_NestedLinkedSourceTests+LinkedSource/Person"
      },
      "children": [
        {
          "node": {
            "referenceType": "HeterogeneousDataSources.Tests.Shared.Image, HeterogeneousDataSources.Tests, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null",
            "linkTargetId": "HeterogeneousDataSources.Tests.PersonLinkedSource/SummaryImage"
          },
          "children": []
        }
      ]
    },
    {
      "node": {
        "referenceType": "HeterogeneousDataSources.Tests.Shared.Image, HeterogeneousDataSources.Tests, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null",
        "linkTargetId": "HeterogeneousDataSources.Tests.ReferenceTree_NestedLinkedSourceTests+LinkedSource/PostImage"
      },
      "children": []
    }
  ]
}